src/git-repository-provider.coffee

```diff
@@ -0,0 +1,80 @@
+fs = require 'fs'
+GitRepository = require './git-repository'
+
+# Checks whether the specified directory has, or has a parent directory that
+# has, a valid .git folder, indicating the root of a Git repository.
+# If found, a Directory that corresponds to the .git folder will be returned.
+# Otherwise, returns `null`.
+#
+# * `directory` {Directory} to explore whether it is part of a Git repository.
+findGitDirectorySync = (directory) ->
+  # TODO: Fix node-pathwatcher/src/directory.coffee so the following methods
+  # can return cached values rather than always returning new objects:
+  # getParent(), getFile(), getSubdirectory().
+  gitDir = directory.getSubdirectory('.git')
+  if directoryExistsSync(gitDir) and isValidGitDirectorySync gitDir
+    gitDir
+  else if directory.isRoot()
+    return null
+  else
+    findGitDirectorySync directory.getParent()
+
+# Returns a boolean indicating whether the specified directory represents a Git
+# repository.
+#
+# * `directory` {Directory} whose base name is `.git`.
+isValidGitDirectorySync = (directory) ->
+  # To decide whether a directory has a valid .git folder, we use
+  # the heuristic adopted by the valid_repository_path() function defined in
+  # node_modules/git-utils/deps/libgit2/src/repository.c.
+  return directoryExistsSync(directory.getSubdirectory('objects')) and
+      directory.getFile('HEAD').exists() and
+      directoryExistsSync(directory.getSubdirectory('refs'))
+
+# Returns a boolean indicating whether the specified directory exists.
+#
+# * `directory` {Directory} to check for existence.
+directoryExistsSync = (directory) ->
+  # TODO: Directory should have its own existsSync() method. Currently, File has
+  # an exists() method, which is synchronous, so it may be tricky to achieve
```

Comment:
Damn. I'm pissed that we missed that! I think we should probably bit the bullet deprecate it post-freeze and favor existsSync to set ourselves up for an async API past 1.0. @kevinsawicki what would you think about that?